### PR TITLE
docs: add documentation about batching in jina

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

**Review by Yongxuanzhang — COMMENTED** [review]
Did you add this in the index?

**Comment by JoanFM** [comment]
> Did you add this in the index?

I realize not, how should I do this

File: chapters/batching.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
+=============================================
+ Understanding Batching in Jina
+=============================================
+
+Some :term:`Executors` in Jina may profit from the capacity to process multiple instances of their input at the same time.
+This can lead to a speed up and higher throughput.
+
+In order to provide this capability, the :term:`Driver` extracts all the needed information from all the :term:`Document` in a :term:`Request`. [see `guide on request size`_].
+
+Then all these instances (one per :term:`Document`) is passed to the :term:`Executor`. Then the :term:`Executor` decides how it wants to
```
Comment (FionnD) [inline chapters/batching.rst:11]: are passed

Follow that, the :term:`Executor` (sounds nice then two sentences with then)

**Review by FionnD — APPROVED** [review]
English and style is good, can't assess the technical aspect

**Comment by JoanFM** [comment]
@Yongxuanzhang can u edit this to put it in the index?

**Comment by Yongxuanzhang** [comment]
> @Yongxuanzhang can u edit this to put it in the index?

Yeah. Sure! I will add it after I read this section. 😄

File: chapters/batching.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
+=============================================
+ Understanding Batching in Jina
+=============================================
+
+Some :term:`Executors` in Jina may profit from the capacity to process multiple instances of their input at the same time.
+This can lead to a speed up and higher throughput.
+
+In order to provide this capability, the :term:`Driver` extracts all the needed information from all the :term:`Document` in a :term:`Request`. [see `guide on request size`_].
```
Comment (Yongxuanzhang) [inline chapters/batching.rst]: Should add a link for `guide on request size`_ maybe? I can add this if needed.

File: chapters/batching.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    .. code-block:: python
+
+        class CapitalizerCrafter(BaseCrafter):
+
+            @batching(batch_size=64)
+            def craft(self, text: Iterable[str], *args, **kwargs):
+                pass
+
+- Make it an attribute of the :term:`Executor` explicitly:
+
```
Comment (Yongxuanzhang) [inline chapters/batching.rst]: Need to remove this blank line to make it consistent with the other 2 ways

File: chapters/batching.rst [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            batch_size: 64
+
+
+When do I need to set these decorators?
+------------
+
+Currently, there are 4 classes of our :term:`Executor` that receive input from :term:`Driver` in batches, and therefore, all the
+classes of these families need to make sure that their core methods are decorated with either `single` or `batching`.
+
+These :term:`Executor` are:
```
Comment (Yongxuanzhang) [inline chapters/batching.rst:134]: There are 5 classes but you mentioned there are 4? 
And also they are not shown properly

File: chapters/batching.rst (hunk 3 above)
Comment (JoanFM) [reply chapters/batching.rst]: Do what u need to do

File: chapters/batching.rst (hunk 4 above)
Comment (JoanFM) [reply chapters/batching.rst:134]: I have no clue about formatting to be honest

File: chapters/batching.rst [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+consume these instances (in batches or one by one).
+
+Batching decorators
+--------------------
+To control how an :term:`Executor` consumes its incoming inputs, Jina provides a set of decorator functions.
+
+single and single_multi_input
+------------------------------
+The single decorator ensures that even when the function is called with a set of instances, they will be processed one by one by the executor.
+This decorator is useful when the executor won't have benefits of processing inputs in batches, and allows the executor to keep a clean interface.
```
Comment (Yongxuanzhang) [inline chapters/batching.rst:28]: Sorry, I'm still not sure I understand the benefits.
Does this(single decorator) means that we first have 1 batch of docs, then we can craft them into 1 by 1 list of docs?
What is the case when _the executor won't have benefits of processing inputs in batches_? What does it mean to keep a clean interface? Do we have an example if it's not a clean interface?

Comment (JoanFM) [reply chapters/batching.rst:28]: So see the example there, it is not the same to have a simple instance ( a str ) than having a list of strings of size 1

File: chapters/batching.rst [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+In the case where this decorator is used, the method needs to be implemented considering a single input.
+
+The single_multi_input is the equivalent decorator required when multiple input arguments are required by the :term:`Executor` method.
+
+.. highlight:: python
+.. code-block:: python
+
+    from jina.executors.crafters import BaseCrafter
+    from jina.executors.decorators import single
+    class CapitalizerCrafter(BaseCrafter):
```
Comment (bwanglzu) [inline chapters/batching.rst:40]:
```suggestion

    class CapitalizerCrafter(BaseCrafter):
```

Comment (bwanglzu) [reply chapters/batching.rst:40]: image

File: chapters/batching.rst [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+
+batching and batching_multi_input
+----------------------------------
+The batching decorator ensures that data is processed in batches of a given size. This is useful when an :term:`Executor` can benefit
+from processing multiple inputs at a time.
+
+In the case where this decorator is used, the method needs to be implemented considering input comes in batches.
+
+The batching_multi_input is the equivalent decorator required when multiple input arguments are required by the :term:`Executor` method.
```
Comment (bwanglzu) [inline chapters/batching.rst]:
```suggestion
The :meth:`batching_multi_input` is the equivalent decorator required when multiple input arguments are required by the :term:`Executor` method.
```

**Review by bwanglzu — CHANGES_REQUESTED** [review]
Looks really nice, one suggestion, can we add a `table of content` section?

**Comment by JoanFM** [comment]
> Looks really nice, one suggestion, can we add a `table of content` section?

Yes, how would we do it?

**Comment by bwanglzu** [comment]
@JoanFM , it will generate the table of contents automatically

```
.. meta::
   :description: some description
   :keywords: some keywords for seo purpose

.. contents:: Table of Contents
    :depth: 2
```

File: chapters/batching.rst (hunk 2 above)
Comment (JoanFM) [reply chapters/batching.rst]: Can u do so? I thought the link was added

Comment (Yongxuanzhang) [reply chapters/batching.rst]: Yes, I added. No worries.

**Review by bwanglzu — APPROVED** [review]
LGTM!

**Comment by Kelton8Z** [comment]
The doc is really helpful

**Pull request merged**